1st iteration: MVP

bgzzz opened the pull request.

Commit:
init(solution): add mvp go-schedule solution

- add common package with minor definitions and functions
- add management-node folder with management node src
- add schedctl folder with schedctl src
- add wrp package with worker node rpc primitives
- add worker-node folder with worker node src

Commit:
test(unit): add simple tests for utility functions

- add unit tests for management node utility funcions
- add unit tests for common package
- add unit tests for schedctl utility functions

- change some timer behaviour on the basis of test output

Commit:
refactor(config): refactored config usage

- common
- for schedctl
- management node
- worker node

Commit:
fix(schedule): concurrent rsp issue

- fixed concurrent task state change
- style issues in wrpc

Commit:
feat(infrastructure): add make files and docker related files

- add docker file to the solution
- add make files to specific part of the solution
- add common make file
- remove useless println

Commit:
docs(readme): add main readme

- how to
- arch and principles
- todo

Commit:
fix(deps): fixed deps warning issue

- update the lock to fetch proper version

Commit:
fix(del): deleted commented file

- deleted useless file

r0mant commented on common/logging.go:
```diff
@@ -0,0 +1,46 @@
+
+	lvl, err := log.ParseLevel(cfg.LogLvl)
+	if err != nil {
+		return err
```
JFYI, we use https://github.com/gravitational/trace when returning errors everywhere so it is possible to print a call stack on error:

```go
return trace.Wrap(err)
```

bgzzz replied:
Done

r0mant commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+	// silence periods during bi-directional streaming
+	// session. Server terminates stream if SilenceTimeout
+	// expired
+	SilenceTimeout int `yaml:"silence_timeout"`
```
Any reason this is (and other `*Timeout` fields here) not a `time.Duration`?

bgzzz replied:
Done

r0mant commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+
+	// Etcd related configuration
+	// EtcdAddress is the address of db
+	EtcdAddress string `yaml:"etcd_address"`
```
Is this a single etcd node address, or can this accept multiple nodes in some format, e.g comma-separated?

bgzzz replied:
By design of the first step, it was only one node of the etcd. I will change it in future.

bgzzz replied:
change it such that there is a reconnection timer and worker tries to reconnect if it loses the connection

r0mant commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+
+// Config is global config variable that holds server
+// configuration
+var Config *ServerConfig
```
Why make it a global variable, rather than, say, have it in main function and initialize server with it?

bgzzz replied:
Done

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+
+	err := mn.SetToDb(wn, EtcdWorkerPrefix+wn.WN.Id)
+	if err != nil {
+		log.Error(err.Error())
```
As a general rule of thumb, all errors should be propagated to the top-level caller and logged there once, instead of logging each error individually.

```go
if err != nil {
  return trace.Wrap(err)
}
```

bgzzz replied:
done

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+// AddWorkerNode thread safely adds worker node to worker pool and etcd
+// errors in case of data inconsistency
+func (mn *ManagementNode) AddWorkerNode(wn *wrpc.WorkerRPCClient) error {
+	mn.workerNodePoolMtx.Lock()
```
It is generally not a very good idea to grab a lock on an I/O operation (database call in this case). Any way to avoid this?

bgzzz replied:
Have change it for add/rm worker. Now, no mutexes over db IO operations.

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(),
```
It's generally a good idea to have context passed by the caller so it can cancel it / etc.

bgzzz replied:
Have done with context passing from top to bottom

r0mant commented on schedctl/calls.go:
```diff
@@ -0,0 +1,138 @@
+// were registered to stdout
+func ListTasks() error {
+	return runRPC(func(client pb.SchedulerClient) error {
+		ctx, cancel := context.WithTimeout(context.Background(),
```
Pass context from the outside (same for other methods here).

r0mant commented on worker-node/main.go:
```diff
@@ -0,0 +1,49 @@
+
+	// Set up a connection to the server.
+
+	conn, err := grpc.Dial(Config.ServerAddress, grpc.WithInsecure())
```
How is trust established between worker nodes and management node?

r0mant commented on wrpc/client.go:
```diff
@@ -0,0 +1,199 @@
+	// if request Id set by uuid string
+	// for now uuid equals to any string with len 35
+	// we do not create req id
+	if len(req.Id) < 35 {
```
That's a bit weird, wouldn't `if req.Id == ""` work? :)

r0mant commented on wrpc/server.go:
```diff
@@ -0,0 +1,156 @@
+			}
+		}
+		//this one is done to prevent go routines leaking
+	case <-t.C:
```
Timer will be leaked if this fires before the case above.

r0mant commented on wrpc/worker.go:
```diff
@@ -0,0 +1,149 @@
+	go func() {
+		for {
+			select {
+			case <-w.stopSender:
```
I wonder, why not use contexts instead of "stop" channels?

r0mant commented on wrpc/worker.go:
```diff
@@ -0,0 +1,149 @@
+	}()
+	t := time.NewTimer(w.silenceTimeout * time.Second)
+	select {
+	case <-t.C:
```
Timer will be leaked here. I'd generally do `defer t.Stop()` right after timer is created to avoid this.

r0mant commented on wrpc/worker.go:
```diff
@@ -0,0 +1,149 @@
+		msg, err := w.RxWithTimeout()
+		if err != nil {
+			log.Error(err.Error())
+			return err
```
What happens when "init loop" exist with error? Does the worker process exit too?

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+		clientv3.WithPrefix(),
+	}
+
+	eCtx, cancel := context.WithTimeout(context.Background(),
```
Why not use context passed to this method as a parent context here?

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+		clientv3.WithPrefix(),
+	}
+
+	eCtx, cancel := context.WithTimeout(context.Background(),
```
Use context passed to this methods as parent?

bgzzz replied:
done

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+
+		// RR worker id calculation
+		if Config.SchedulerAlgo == SchedAlgoRR {
+			workerId = workers[int(math.
```
If you schedule tasks one-by-one, then using this algorithm it looks like they all will be scheduled on the first worker node? Unless I'm missing something.

bgzzz replied:
i is index going from 0 to len(workers)
let's assume len(workers) = 10
math.Mod(1,10) = 1, math.Mod(2,10) = 2 .... math.Mod(10,10) = 0 , math.Mod(11,10) = 1

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+
+	// channel for connecting onRsp handlers
+	// to the current go routine
+	messanger := make(chan msg, 1)
```
Nit: mess**e**nger?

bgzzz replied:
refactored it

r0mant commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+			task.State = common.TaskStateDead
+			err := md.SetToDb(task, EtcdTaskPrefix+task.Id)
+			if err != nil {
+				log.Error(err.Error())
```
Just to reiterate something I mentioned above, instead of logging each error every time, just propagate them to the top level and log there, so all `log.Error(err.Error())` aren't necessary (and use `trace.Wrap(err)`)

bgzzz replied:
done

r0mant commented on schedctl/config.go:
```diff
@@ -0,0 +1,40 @@
+	"io/ioutil"
+)
+
+var Config *SchedCtlConfig
```
Same question as in the "management node" part - any good reason this is a global?

bgzzz replied:
done

r0mant reviewed (commented):
First pass.

knisbet commented on common/task.go:
```diff
@@ -0,0 +1,28 @@
+	// TaskStatePending is the state of the task when the
+	// Task is registered in db but has not sent
+	// to worker node
+	TaskStatePending = "pending"
```
Question: I'm curious about why you're defining the states as globals, instead of defining as part of the .proto files as an enum?

bgzzz replied:
Agree about the state. Will try to keep in in DB only. This was the design mistake from my side. It seems ok (separated state handing) with one management node but will be a problem in HA mode

knisbet commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+	// EtcdAddress is the address of db
+	EtcdAddress string `yaml:"etcd_address"`
+	// DialTimeout is timeout used for etcd querying
+	DialTimeout int `yaml:"dial_timeout"`
```
Nitpick: Since this appears to be specific to etcd, for variable naming I would consider including etcd in the name.

bgzzz replied:
done

knisbet commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+// cfgValidate validates some config fields
+func cfgValidate(cfg *ServerConfig) error {
+
+	if cfg.SchedulerAlgo != SchedAlgoRR &&
```
Nitpick: I would generally apply this as a switch statement, just makes it clearer/easier to maintain down the road if new scheduler algorithms get added.

bgzzz replied:
will do it in future

knisbet commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+	// cfg.PingTimer has to be <= cfg.SilenceTimeout
+	// due to async architecture
+	if cfg.PingTimer > cfg.SilenceTimeout {
+		cfg.PingTimer = cfg.SilenceTimeout
```
Based on the description of SilenceTimeout/PingTimer, I don't think it's safe for both timers to be equal. 

If there is a delay in the ping, because of load or network latency/packet loss, couldn't that cause the ping to get slightly delayed, and then the Silence timer to get triggered?

bgzzz replied:
Agree

knisbet commented on management-node/config.go:
```diff
@@ -0,0 +1,97 @@
+	// cfg.PingTimer has to be <= cfg.SilenceTimeout
+	// due to async architecture
+	if cfg.PingTimer > cfg.SilenceTimeout {
+		cfg.PingTimer = cfg.SilenceTimeout
```
Question: When validating the configuration, why the choice to edit the configuration instead of returning a configuration error?

bgzzz replied:
done

knisbet commented on management-node/config_test.go:
```diff
@@ -0,0 +1,77 @@
+	"github.com/bgzzz/go-schedule/common"
+)
+
+func TestConfigValidate(t *testing.T) {
```
Nitpick: it looks like these tests are close enough to each other, they could reasonably be converted to a table driven test at some point. https://github.com/golang/go/wiki/TableDrivenTests

bgzzz replied:
ok, will do it in future

russjones commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+	mn.workerNodePoolMtx.Lock()
+	defer mn.workerNodePoolMtx.Unlock()
+
+	if _, ok := mn.workerNodePool[wn.WN.Id]; ok {
```
Why not just maintain all this information on etcd? Why maintain it in-memory as well?

bgzzz replied:
refactored it such that all the state is in db

russjones commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+}
+
+// SetToDb stores subj to etcd by id
+func (md *ManagementNode) SetToDb(subj interface{}, id string) error {
```
I've move all this code to another package right now `ManagementNode` node mixes business logic and data.

bgzzz replied:
ok, will do it in future

russjones commented on schedctl/calls.go:
```diff
@@ -0,0 +1,138 @@
+// it connects to management node and runs specific call
+func runRPC(call func(client pb.SchedulerClient) error) error {
+
+	conn, err := grpc.Dial(Config.ServerAddress, grpc.WithInsecure())
```
I'd prioritize work on auth over things like the scheduling algorithm.

russjones commented on schedctl/calls.go:
```diff
@@ -0,0 +1,138 @@
+		return err
+	}
+
+	return runRPC(func(client pb.SchedulerClient) error {
```
I'd just make a client and attach it to a struct and then you can use in it any method.

knisbet commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+//		or even changed to stream
+func (md *ManagementNode) GetWorkerList(ctx context.Context, req *pb.DummyReq) (*pb.WorkerNodeList, error) {
+
+	log.Info("GetWorkerList")
```
Nitpick: probably should be a debug level + log the request.

knisbet commented on management-node/server.go:
```diff
@@ -0,0 +1,530 @@
+			workerId = workers[int(math.
+				Mod(float64(i),
+					float64(len(workers))))]
+		} else if Config.SchedulerAlgo == SchedAlgoRand {
```
Nitpick: It might be worth handling an unexpected scheduler here. While on the current iteration I believe it's totally unnecessary, it's one of those defence in depth things, that if a scheduler get's added later, but is missed in this code section, this code will fail / behave unexpectedly if it processes a request for the new scheduler.

russjones commented on wrpc/client.go:
```diff
@@ -0,0 +1,199 @@
+
+// StartPinger setups ticker and sends ping to worker on
+// every tick
+func (wn *WorkerRPCClient) StartPinger(d time.Duration) {
```
I think a more idiomatic pattern would be to just call a function with a context that has a deadline instead of using callbacks like this.

knisbet reviewed (commented):
I think my overall largest concern, is the way state is maintained in memory and etcd. It creates sort of two sources of authority, in most cases reads / work is done just in memory, but is synchronized to etcd. This leaves a lot of room for discrepancies to exist for various reasons, and it looks like you take a lot of long held locks to help protect from internal consistency problems.

An example of this, is the  DeadTimeout in the management node, as far as I can tell, it doesn't get re-created if the management process is restarted. Because etcd is being used for a state storage, I would be left with the impression that the process could restart. 

I'm curious if you don't mind, would you share your thought process on this, and why this route was chosen?

As a second note, can you share in some way, how you would secure this system?

Commit:
style(err): change error handling

- add trace functionality while returning errors
- remove stack trace logging
- add logging only to the highest possible level

- add .gitignore

ERROR HANDLING

Commit:
fix(config): remove global config

- rm global config varaible from each part of the solution

Commit:
fix(worker sync): change mutex operation logic

- remove mutexes on top of io operations for add/rm worker

Commit:
refactor(context): switch timers to context timers

- change all timers to context with timers

Commit:
refactor(schedule): refactored scheduling function

refactor it such that task states are changed with following order:
pending -> set to db wihtout allocated worker
penfing -> set to db with allocated worker
scheduled -> set to db
done/dead -> set to db

schedule gRPC returns right after tasks are set to db with pending state

then in different go routine scheduler is called to allocated and execute tasks

Commit:
test(fixed): fixed previously written tests

- fixed tests
- add function for MN creation
- fix task timers

Commit:
tests(server): add some integrational test for managment node server

- add integrational tests for management node
- add NewObj functions for mn and tasks

Commit:
feat(worker): add reconnect feature to worker node

- add reconnect feature
- add reconnect timeout
